iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{agent_error::ImlAgentError, lustre::search_rootpath};
+use futures::{future::try_join_all, future::Future, stream, StreamExt, TryStreamExt};
+use iml_wire_types::{FidError, FidItem};
+use liblustreapi::LlapiFid;
+use std::path::PathBuf;
+use structopt::clap::arg_enum;
+use tokio::{fs, process::Command};
+
+struct Work {
+    fid: FidItem,
+    src_file: String,
+    dest_file: String,
+}
+
+fn do_rsync<'a>(
+    work_list: &'a [Work],
+) -> impl Future<Output = Result<Vec<FidError>, ImlAgentError>> + 'a {
+    stream::iter(work_list)
+        .map(|work| async move {
+            let output = Command::new("rsync")
+                .arg("-a")
+                .arg("-t")
+                .arg("-X")
+                .arg(&work.src_file)
+                .arg(&work.dest_file)
+                .output()
+                .await?;
+
+            Ok::<_, ImlAgentError>(FidError {
+                fid: work.fid.fid.clone(),
+                data: work.fid.data.clone(),
+                errno: output.status.code().unwrap_or(0) as i16,
+            })
+        })
+        .chunks(10)
+        .map(Ok)
+        .try_fold(vec![], |mut acc, xs| async {
+            let xs = try_join_all(xs).await?;
+
+            acc.extend(xs);
+
+            Ok::<_, ImlAgentError>(acc)
+        })
+}
+
+async fn archive_fids(
+    llapi: LlapiFid,
+    dest_src: &str,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = vec![];
+
+    let mut rsync_list: Vec<Work> = vec![];
+
+    for fid in fid_list {
+        let fid_path = match llapi.fid2path(&fid.fid) {
+            Ok(x) => x,
+            Err(_) => {
+                tracing::error!("llapi: failed on fid {}", &fid.fid);
+                continue;
+            }
+        };
+
+        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
+        let dest_file = format!("{}/{}", &dest_src, &fid_path);
+        let mut dest_dir = PathBuf::from(&dest_file);
+
+        dest_dir.pop();
+        if !dest_dir.is_dir() {
+            fs::create_dir_all(&dest_dir).await?;
+        }
+
+        let md = fs::metadata(&src_file).await?;
+        /* invoking mpifileutils is slow, so only do it for directories
+         * and big files, otherwise just use rsync
+         */
+        if md.is_dir() || (md.len() > 1024 * 1024 * 1024) {
```

Comment:
Would be nice if "too big" was a constant or configurable

Reply:
Done